### PR TITLE
feat(auth): add authentication guide and api references

### PR DESCRIPTION
**Pull request by meskill**

### CODE REVIEW

File: docs/guides/auth.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+- **Protect Sensitive Information** Some fields may contain sensitive information that should not be accessible to unauthenticated users.
+- **Prevent Unauthorized Access** Ensuring that authenticated users can query certain fields helps prevent unauthorized access and potential data breaches.
+- **Customized User Experiences** Authentication allows for personalized queries based on the user's identity.
+
+The authentication could be implemented by different credential validation mechanisms, like:
+
+- **Basic Authentication** is a straightforward authentication scheme that sends base64-encoded usernames and passwords in the HTTP Authorization header with each request. It's simple to implement but requires HTTPS to ensure security due to its lack of encryption.
+- **JWT Authentication** uses digitally signed tokens to authenticate and transmit user information in a compact JSON format, allowing stateless and secure communication between clients and servers. It offers greater flexibility and security, supporting expiration times and custom data embedding within the token itself.
```
Comment (tusharmath) [inline docs/guides/auth.md]: Should have two separate guides about JWT and Basic Auth.

File: docs/guides/auth.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+## What is Authentication?
+
+Authentication in GraphQL refers to the process of verifying the identity of a user before allowing access to specific fields in GraphQL schema. Essentially, it's about controlling who can see or interact with certain pieces of data exposed by GraphQL API. This process is crucial for following reasons:
+
+- **Protect Sensitive Information** Some fields may contain sensitive information that should not be accessible to unauthenticated users.
+- **Prevent Unauthorized Access** Ensuring that authenticated users can query certain fields helps prevent unauthorized access and potential data breaches.
+- **Customized User Experiences** Authentication allows for personalized queries based on the user's identity.
+
+The authentication could be implemented by different credential validation mechanisms, like:
```
Comment (tusharmath) [inline docs/guides/auth.md]: Expand a bit on how it's different from existing auth mechanisms, particularly around field level auth.

File: docs/guides/auth.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+type Post {
+  id: Int!
+  userId: Int!
+  title: String!
+  body: String! @protected
+  user: User @http(path: "/users/{{value.userId}}")
+}
+```
+
+In that case the whole `User` type and `Post.body` are marked as protected and therefore requiring use authentication to resolve its content.
```
Comment (tusharmath) [inline docs/guides/auth.md]: Should show working examples with queries and potential error codes.

File: docs/guides/auth.md (hunk 1 above)
Comment (tusharmath) [reply docs/guides/auth.md]: Have a section for authentication and then inside it have three files — Introduction, JWT and then Basic Auth

Comment (tusharmath) [reply docs/guides/auth.md]: Also need to expand on how auth work for mutations. Provide examples.

File: docs/auth/introduction.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+---
+
+This guide will walk you through entity level authentication in GraphQL and how it could be achieved with Tailcall.
+
+## What is Authentication?
+
+Authentication in GraphQL refers to the process of verifying the identity of a user before allowing access to specific fields in GraphQL schema. Essentially, it's about controlling who can see or interact with certain pieces of data exposed by GraphQL API. This process is crucial for following reasons:
+
+- **Protect Sensitive Information** Some fields may contain sensitive information that should not be accessible to unauthenticated users.
+- **Prevent Unauthorized Access** Ensuring that authenticated users can query certain fields helps prevent unauthorized access and potential data breaches.
```
Comment (amitksingh1490) [inline docs/auth/introduction.md]:
```suggestion
```
can be dropped as this is a case for authorisation not authentication

**Review by amitksingh1490 — APPROVED** [review]

**Pull request merged**